src/k8s/cmd/k8s/k8s_set.go

```diff
@@ -0,0 +1,275 @@
+package k8s
+
+import (
+	"encoding/csv"
+	"fmt"
+	"regexp"
+	"slices"
+	"strconv"
+	"strings"
+
+	api "github.com/canonical/k8s/api/v1"
+	"github.com/canonical/k8s/cmd/k8s/errors"
+	"github.com/spf13/cobra"
+)
+
+var (
+	settingsMap = map[string][]string{
+		"network":       []string{"enabled"},
+		"dns":           []string{"enabled", "upstream-nameservers", "cluster-domain", "service-ip"},
+		"gateway":       []string{"enabled"},
+		"ingress":       []string{"enabled", "default-tls-secret", "enable-proxy-protocol"},
+		"local-storage": []string{"enabled", "local-path", "reclaim-policy", "set-default"},
+		"load-balancer": []string{
+			"enabled",
+			"cidrs",
+			"l2-mode",
+			"l2-interfaces",
+			"bgp-mode",
+			"bgp-local-asn",
+			"bgp-peer-address",
+			"bgp-peer-port",
+			"bgp-peer-asn",
+		},
+		"metrics-server": []string{"enabled"},
+	}
+)
+
+func newSetCmd() *cobra.Command {
+	setCmd := &cobra.Command{
+		Use:               "set <functionality.key=value>...",
+		Short:             "Set functionality configuration",
+		PersistentPreRunE: chainPreRunHooks(hookSetupClient),
+		Args: func(cmd *cobra.Command, args []string) error {
+			// Optionally run one of the validators provided by cobra
+			if err := cobra.ArbitraryArgs(cmd, args); err != nil {
+				return err
+			}
+
+			var re = regexp.MustCompile(`([^\. ]+)\.([^\. ]+)=(.+)`)
+			for _, arg := range args {
+				matches := re.FindStringSubmatch(arg)
+				if len(matches) < 4 {
+					return fmt.Errorf("invalid argument format: %s", arg)
+				}
+
+				functionality := matches[1]
+
+				settings, ok := settingsMap[functionality]
+				if !ok {
+					return fmt.Errorf("invalid functionality: %s", arg)
+				}
+
+				key := matches[2]
+				if !slices.Contains(settings, key) {
+					return fmt.Errorf("invalid key: %s", arg)
+				}
+			}
+
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			defer errors.Transform(&err, nil)
+
+			config := api.UserFacingClusterConfig{}
+
+			settingsMap := map[string]map[string]string{}
+
+			for _, arg := range args {
+				splitArg := strings.Split(arg, "=")
+				splitField := strings.Split(splitArg[0], ".")
+				functionality := splitField[0]
+				key := splitField[1]
+				value := splitArg[1]
+
+				_, ok := settingsMap[functionality]
+
+				if !ok {
+					settingsMap[functionality] = map[string]string{}
+				}
+
+				settingsMap[functionality][key] = value
+			}
+
+			for functionality, configs := range settingsMap {
```

Comment:
Let's use a flat map instead, it should be easier to follow here, and possibly extend in the future.

E.g.

```
for _, arg := range args {
  key, value = splitKeyAndValue(arg)
  switch key {
  case "network.enabled":
    // ....
  default:
    // unknown config
  }
}
```